esp/services/ws_fs/ws_fsService.cpp

```diff
@@ -1881,49 +1881,132 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
+            }
+            else
+            {
+                SocketEndpoint hostInPathEp;
+                hostInPathEp.set(hostInPath);
+                if (!srcHostEp.ipequals(hostInPathEp))
+                {
+                    VStringBuffer msg("The network address %s defined in the SourcePath does not match with the network address %s defined in SourceIP. ",
+                        hostInPath, sourceHostReq.str());
+                    msg.append("A dropzone name specified in the SourcePlane is preferred. The host/IP should not be contained in the SourcePath.");
+                    throw makeStringException(ECLWATCH_INVALID_INPUT, msg.str());
+                }
+            }
+            fileReq = localPath.str();
+        }
+
+        if (containsRelPaths(fileReq)) //Detect a path like: a/../../../f
+            throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Invalid path %s", fileReq);
 
         if (!isEmptyString(srcPlane))
         {
-            Owned<IPropertyTree> dropZone = getDropZonePlane(srcPlane);
-            if (!dropZone)
-                throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Unknown landing zone: %s", srcPlane);
-            const char * dropZonePlanePath = dropZone->queryProp("@prefix");
-            if (isAbsolutePath(sourcePath))
+            if (isAbsolutePath(fileReq))
             {
-                if (!startsWith(sourcePath, dropZonePlanePath))
-                    throw makeStringException(ECLWATCH_INVALID_INPUT, "Invalid source path");
+                //Check if the fileReq is valid.
+                if (!startsWith(fileReq, dropZonePlanePath))
+                    throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Invalid source path %s.", fileReq);
             }
             else
             {
-                StringBuffer s(sourcePath);
-                sourcePath.set(dropZonePlanePath);
-                addNonEmptyPathSepChar(sourcePath);
-                sourcePath.append(s);
+                //Insert the @prefix in the front of the file
+                localPath.set(dropZonePlanePath);
+                addNonEmptyPathSepChar(localPath);
+                localPath.append(fileReq);
+                fileReq = localPath.str();
             }
-            getDropZoneHost(srcPlane, dropZone, sourceIPReq);
-            sourcePlaneReq.append(srcPlane);
         }
         else
         {
-            sourceIPReq.set(srcIP).trim();
-            if (sourceIPReq.isEmpty())
-                throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Source network IP not specified.");
-            Owned<IPropertyTree> plane = getAndValidateDropZone(sourcePath, sourceIPReq);
-            sourcePlaneReq.append(plane->queryProp("@name"));
+            if (firstFileReq)
+            {
+                firstFileReq = false;
+
+                //Use the fileReq and sourceHostReq to find out dropzone plane.
+                dropZone.setown(getAndValidateDropZone(fileReq, sourceHostReq));
```

Comment:
it feels this should proceed the if (isAbsolutePath(fileReq)) block (code between lines 1965-1978).
It also might make this method clear, if that code were in a separate (static) helper function (can still be within ws_fsService.cpp).

Reply:
If this proceeds the if (isAbsolutePath(fileReq)) block (code between lines 1965-1978), the getAndValidateDropZone() will be called when the Source Plan is not empty.
I am not sure we want to call the getAndValidateDropZone() when the Source Plane is not empty. When the Source Plane is not empty, the sourceHostReq has been validated and the fileReq will be validated inside the  if (isAbsolutePath(fileReq)) block.